mesonbuild/minstall.py

```diff
@@ -556,13 +557,36 @@ def do_install(self, datafilename: str) -> None:
                     self.log('Preserved {} unchanged files, see {} for the full list'
                              .format(self.preserved_file_count, os.path.normpath(self.lf.name)))
         except PermissionError:
-            if shutil.which('pkexec') is not None and 'PKEXEC_UID' not in os.environ and destdir == '':
-                print('Installation failed due to insufficient permissions.')
-                print('Attempting to use polkit to gain elevated privileges...')
-                os.execlp('pkexec', 'pkexec', sys.executable, main_file, *sys.argv[1:],
-                          '-C', os.getcwd())
-            else:
+            if is_windows() or destdir != '' or not os.isatty(sys.stdout.fileno()) or not os.isatty(sys.stderr.fileno()):
+                # can't elevate to root except in an interactive unix environment *and* when not doing a destdir install
                 raise
+            rootcmd = os.environ.get('MESON_ROOT_CMD') or shutil.which('sudo') or shutil.which('doas')
+            pkexec = shutil.which('pkexec')
+            if rootcmd is None and pkexec is not None and 'PKEXEC_UID' not in os.environ:
```

Comment:
Why is sudo preferred over pkexec? AFAIK systemctl uses polkit, that makes it more consistent with modern CLI. I think polkit will always ask for password so we don't need Y/N selection in that case.

Reply:
Also should document at least the new MESON_ROOT_CMD env.

Reply:
"Modern CLI"? I'm genuinely unsure what you're referring to.

Anyway, pkexec is an ugly, ugly hack. Systemd doesn't use it, it uses the polkit dbus API to do privileged operations via a client/server architecture IIRC (so you can mix privileged and unprivileged actions ) -- like most polkit software does, and, like polkit is intended to be used. This matters because IIRC polkit tracks which process received certain permissions and doesn't ask for a password again. Highly relevant... to GUI programs you don't want to run as root. And pkexec does none of this! It just runs the entire program as root.

Anyway, I could go on and on and on about why I don't like polkit for CLI programs, but I don't think this is really the right place for that. Suffice to say that I'm dead set against it.

But that's fine. If people really want to use it, they can -- just not by default if a better method exists.

Reply:
TBH, I like pkexec only because it pops a GUI to prompt for password, a more standard UX for desktop usage at least. But I agree there seems to have tones of downsides, so not worth it. Let's do it your way then, works perfectly indeed.